src/d3a/blockchain/contracts/ink/lib.rs

```diff
@@ -0,0 +1,203 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract(version = "0.1.0")]
+mod erc20 {
+    use ink_core::storage;
+
+    #[ink(storage)]
+    struct Erc20 {
+        /// The total supply.
+        total_supply: storage::Value<Balance>,
+        /// The balance of each user.
+        balances: storage::HashMap<AccountId, Balance>,
+        /// Approval spender on behalf of the message's sender.
+        allowances: storage::HashMap<(AccountId, AccountId), Balance>,
+        /// The balance of each user.
+        trades: storage::HashMap<Hash, (AccountId, AccountId, Balance, i8)>,
+    }
+
+    #[ink(event)]
+    struct Transfer {
+        #[ink(topic)]
+        from: Option<AccountId>,
+        #[ink(topic)]
+        to: Option<AccountId>,
+        value: Balance,
+    }
+
+    #[ink(event)]
+    struct Approval {
+        #[ink(topic)]
+        owner: Option<AccountId>,
+        #[ink(topic)]
+        spender: Option<AccountId>,
+        #[ink(topic)]
+        value: Balance,
+    }
```

Comment:
Assuming the target of this contract to only track trade, remnants of token transfer could safely be removed.

Reply:
For now, we decided that we will leave the ERC20 standard plus the trade structures as extension